### PR TITLE
Fix sort order aware file group parallelization 

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/8505~

## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/8451

## Rationale for this change

Repatitioning data for pre-sorted listing tables can sometimes result in incorrect results. See descriptions on https://github.com/apache/arrow-datafusion/issues/8451 and in the comments in this PR for details

## What changes are included in this PR?
1. Move the code / tests for redistributing files amongst groups to its own module
2. Add code + tests  to handle redistributing files and preserving sort order

## Are these changes tested?
Yes, new unit tests and end to end coverage (updates to https://github.com/apache/arrow-datafusion/pull/8505) 

## Are there any user-facing changes?
Correct answers with pre-sorted data

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/enforce_distribution.rs [hunk 1]
```diff
@@ -4585,15 +4584,11 @@ pub(crate) mod tests {
...
             // Since at the start of the rule ordering requirement is satisfied
             // EnforceDistribution rule satisfy this requirement also.
-            // ordering is re-satisfied by introduction of SortExec.
-            "SortExec: expr=[a@0 ASC]",
+            "SortRequiredExec: [a@0 ASC]",
             "FilterExec: c@2 = 0",
-            // ordering is lost here
-            "RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=2",
-            "ParquetExec: file_groups={2 groups: [[x], [y]]}, projection=[a, b, c, d, e], output_ordering=[a@0 ASC]",
+            "ParquetExec: file_groups={10 groups: [[x:0..20], [y:0..20], [x:20..40], [y:20..40], [x:40..60], [y:40..60], [x:60..80], [y:60..80], [x:80..100], [y:80..100]]}, projection=[a, b, c, d, e], output_ordering=[a@0 ASC]",
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_distribution.rs:4591]: The file is now divided into groups that preserve its order and thus no resort is needed

File: datafusion/sqllogictest/test_files/repartition_scan.slt [hunk 2]
```diff
@@ -118,7 +118,7 @@ physical_plan
...
+------ParquetExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/repartition_scan/parquet_table/1.parquet:0..197], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/repartition_scan/parquet_table/2.parquet:0..201], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/repartition_scan/parquet_table/2.parquet:201..403], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/repartition_scan/parquet_table/1.parquet:197..394]]}, projection=[column1], output_ordering=[column1@0 ASC NULLS LAST], predicate=column1@0 != 42, pruning_predicate=column1_min@0 != 42 OR 42 != column1_max@1
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/repartition_scan.slt:121]: This shows the test added in https://github.com/apache/arrow-datafusion/pull/8505 is now fixed (the two files are not intermixed)

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+
+//! Logic for managing groups of [`PartitionedFile`]s in DataFusion
+
+use crate::datasource::listing::{FileRange, PartitionedFile};
+use itertools::Itertools;
+use std::cmp::min;
+use std::collections::BinaryHeap;
+use std::iter::repeat_with;
+
+/// Repartition input files into `target_partitions` partitions, if total file size exceed
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:26]: I added a bunch of comments trying to clarify what this code was supposed to be doing

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+///                                    │ │  File B (1MB)   │ │
+///                                      │                 │
+///                                    │ └─────────────────┘ │
+///                                     ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+///
+///                                    If target_partitions = 4,
+///                                      divides into 4 groups
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct FileGroupPartitioner {
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:125]: The API is new, but the file distribution algorithm is the same for unordered inputs

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+        }
+    }
+
+    /// Evenly repartition files across partitions by size, ignoring any
+    /// existing grouping / ordering
+    fn repartition_evenly_by_size(
+        &self,
+        file_groups: &[Vec<PartitionedFile>],
+    ) -> Option<Vec<Vec<PartitionedFile>>> {
+        let target_partitions = self.target_partitions;
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:208]: This is the old algorithm, unmodified

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+
+        Some(repartitioned_files)
+    }
+
+    /// Redistribute file groups across size preserving order
+    fn repartition_preserving_order(
+        &self,
+        file_groups: &[Vec<PartitionedFile>],
+    ) -> Option<Vec<Vec<PartitionedFile>>> {
+        // Can't repartition and preserve order if there are more groups
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:272]: This is a new algorithm introduced for this PR

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 7]
```diff
@@ -0,0 +1,826 @@
...
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.range_size().cmp(&other.range_size())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    /// Empty file won't get partitioned
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:393]: The first set of tests are the original tests, though I refactored them so they didn't rely on ParquetExec

File: datafusion/core/src/datasource/physical_plan/file_groups.rs [hunk 8]
```diff
@@ -0,0 +1,826 @@
...
+
+        let partitioner = FileGroupPartitioner::new()
+            .with_target_partitions(65)
+            .with_repartition_file_min_size(500);
+
+        assert_partitioned_files(None, repartition_test(partitioner, empty_partition))
+    }
+
+    #[test]
+    fn repartition_ordered_no_action_too_few_partitions() {
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_groups.rs:582]: New tests start here

File: datafusion/core/src/datasource/physical_plan/file_scan_config.rs [hunk 9]
```diff
@@ -176,79 +171,17 @@ impl FileScanConfig {
...
-    /// do actual partition on specific data source type. (e.g. `CsvOpener` will only read lines
-    /// overlap with byte range but also handle boundaries to ensure all lines will be read exactly once)
+    #[allow(missing_docs)]
+    #[deprecated(since = "33.0.0", note = "Use SessionContext::new_with_config")]
     pub fn repartition_file_groups(
         file_groups: Vec<Vec<PartitionedFile>>,
         target_partitions: usize,
         repartition_file_min_size: usize,
     ) -> Option<Vec<Vec<PartitionedFile>>> {
-        let flattened_files = file_groups.iter().flatten().collect::<Vec<_>>();
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/file_scan_config.rs:191]: refactored into datafusion/core/src/datasource/physical_plan/file_groups.rs

File: datafusion/core/src/datasource/physical_plan/mod.rs [hunk 10]
```diff
@@ -809,345 +810,4 @@ mod tests {
             extensions: None,
         }
     }
-
-    /// Unit tests for `repartition_file_groups()`
-    #[cfg(feature = "parquet")]
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/mod.rs:814]: These tests were moved and refactored into datafusion/core/src/datasource/physical_plan/file_groups.rs

File: datafusion/core/src/physical_optimizer/enforce_distribution.rs [hunk 11]
```diff
@@ -3862,6 +3863,56 @@ pub(crate) mod tests {
         Ok(())
     }
 
+    #[test]
+    fn parallelization_multiple_files() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_distribution.rs:3867]: This test fails on main

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
Thank you for the review @Dandandan

**Pull request merged**